### PR TITLE
[ICIoU]: Improved CIoU loss

### PR DESCRIPTION
**Pull request by nrupatunga**
# Improved CIoU loss

This modification is based on the paper: `ICIoU: Improved Loss Based on Complete Intersection Over Union for Bounding Box Regression` [Link](https://www.researchgate.net/publication/353468751_ICIoU_Improved_Loss_Based_on_Complete_Intersection_Over_Union_for_Bounding_Box_Regression)

In CIoU loss, the loss function considers

1. IOU
2. Distance between the centers
3. Aspect ratio of the dimensions (w, h) of the boxes

Aspect ratio part of CIoU loss:
ciou

if `(Wgt/Hgt)=(W/H)`, CIoU loss degenerates to DIoU loss. In order to address this issue, this  part of the loss function is modified to 

ICIOU

Since I do not have the resources to run the experiment, I couldn't run the experiments to see if there are any improvements. Let me know what's the best way to do it? 

Thank you!

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Introduced ICIoU (Integrating Complete Intersection over Union) in YOLOv5's loss and metrics calculations.

### 📊 Key Changes
- 🧮 Modified the bbox_iou function to include an `ICIoU` boolean parameter.
- 🤖 Changed the loss calculation by using ICIoU instead of CIoU.
- ✅ Added conditional branches in the `bbox_iou` function for calculating box sizes and union area, specifically if ICIoU is enabled.
- 📐 Implemented the ICIoU formula by adding an additional if-statement within the existing IoU types' logic.

### 🎯 Purpose & Impact
- 🎯 **Purpose**: To improve the accuracy of bounding box predictions by using a more comprehensive IoU metric during training.
- 🚀 **Impact**: Users may see enhanced model precision and better performance, especially in scenarios where accurately fitting bounding boxes is critical.

### CODE REVIEW

**Comment by glenn-jocher** [comment]
@nrupatunga thanks for the PR! I'll run a VOC comparison. Should be done in a few hours.

**Comment by nrupatunga** [comment]
Thank you @glenn-jocher :)

**Comment by glenn-jocher** [comment]
Test reproduction code:
```python
# Setup
!git clone https://github.com/ultralytics/yolov5 yolov5  # clone
!git clone https://github.com/nrupatunga/yolov5 yolov5-pr -b ICIOU  # clone
%pip install -qr yolov5/requirements.txt wandb  # install

# Weights & Biases  (optional)
import wandb
wandb.login()

# Train
%cd yolov5
!python train.py --batch 64 --weights yolov5s6.pt --data VOC.yaml --epochs 50 --cache --img 512 --nosave --hyp hyp.VOC.yaml --project PR --name YOLOv5s6-master
%cd ../yolov5-pr
!python train.py --batch 64 --weights yolov5s6.pt --data VOC.yaml --epochs 50 --cache --img 512 --nosave --hyp hyp.VOC.yaml --project PR --name YOLOv5s6-pr
```
Results at https://wandb.ai/glenn-jocher/PR

**Comment by nrupatunga** [comment]
@glenn-jocher Curious on how do you infer the behavior of training with CIoU and ICIoU loss (YOLOv5s6-master vs YOLOv5s6-pr), both seem to have similar kinds of metric curves?

Can we say the distribution of predicted boxes during training is such that, `wgt/hgt=w/h`, do now seem to happen?

**Comment by glenn-jocher** [comment]
@nrupatunga I trained VOC for 50 epochs (finetuned from yolov5s6.pt) and could not see any differences, so I did a second comparison training VOC for 300 epochs from scratch. This second training shows better results with master at the end of training, and lower box loss in particular:

https://wandb.ai/glenn-jocher/PR
Screenshot 2022-03-15 at 10 32 30

**Comment by nrupatunga** [comment]
Thank you for running experiments with these changes @glenn-jocher This modification, does not seem to add much gain, and box loss is indeed higher 🤔

**Comment by glenn-jocher** [comment]
@nrupatunga it seems like the proposed feature did not improve the results, but thanks for implementing the PR in any case! It's good to experiment with new features to understand what works well and what doesn't work well.

**Comment by nrupatunga** [comment]
agreed! Thank you for assisting @glenn-jocher

**Comment by nrupatunga** [comment]
closing PR as the changes didn't improve the results!
